---
title: "scRNA-Seq Lund AxD"
author: "Pavel"
date: "16 4 2021"
output: html_document
---

Load required package
```{r, message=FALSE, warning=FALSE, paged.print=TRUE}
suppressMessages(library("scater"))
suppressMessages(library("Matrix"))
suppressMessages(library("NormExpression"))
suppressMessages(library("DropletUtils"))
suppressMessages(library("pheatmap"))
suppressMessages(library("RColorBrewer"))
suppressMessages(library("ggplot2"))
suppressMessages(library("Seurat"))
suppressMessages(library("sctransform"))
suppressMessages(library("DT"))
suppressMessages(library("scales"))
suppressMessages(library("ggrepel"))
suppressMessages(library("dplyr"))
suppressMessages(library("SeuratWrappers"))
suppressMessages(library("scmap"))
suppressMessages(library("SoupX"))
suppressMessages(library("reshape2"))
suppressMessages(library("clustree"))
suppressMessages(library("R.utils"))
suppressMessages(library("Nebulosa"))
suppressMessages(library("monocle3"))
suppressMessages(library("fgsea"))
suppressMessages(library("clusterProfiler"))
suppressMessages(library("org.Hs.eg.db"))
```

```{r}
samples<-read.table("../samples.csv",header=TRUE,sep=",")
```

```{r, include = FALSE}
STAR <- readRDS("STAR_raw.rds")
seurat_final <- readRDS("seurat_final_cell_cylce.rds")
```


```{r eval=FALSE, warning=FALSE}
STAR_merged <- NULL
STAR_merged[["RNA"]] <- cbind(STAR[[1]][["RNA"]],STAR[[2]][["RNA"]],STAR[[3]][["RNA"]],STAR[[4]][["RNA"]],STAR[[5]][["RNA"]],STAR[[6]][["RNA"]])
counts <- STAR_merged[["RNA"]][,colnames(STAR_merged[["RNA"]]) %in% colnames(seurat_final)]
```


```{r eval=FALSE, warning=FALSE}
colData <- seurat_final@meta.data[,c(4:8,14:16,21)]
seurat_all <- CreateSeuratObject(counts = counts, project = "scRNA_OGD_AxD", assay = "RNA", meta.data = colData)
```

```{r, eval = FALSE}
rm(counts, seurat_final, STAR, colData, STAR_merged)
saveRDS(seurat_all,"seurat_all.rds")
```


```{r, eval = TRUE}
seurat_all <- readRDS("seurat_all.rds")
```

```{r, eval = TRUE}
seurat_all[["percent.mt"]] <- PercentageFeatureSet(seurat_all, pattern = "^MT-")
seurat_all[["percent.rib"]] <- PercentageFeatureSet(seurat_all, pattern = "^RPL|^RPS")
```


```{r, fig.width = 4, fig.height = 10}
gridExtra::grid.arrange(VlnPlot(seurat_all, features = c("nFeature_RNA"), pt.size = 0, group.by = "cell_type_1") + scale_y_log10() + NoLegend() + geom_hline(yintercept = 2000) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)),
          VlnPlot(seurat_all, features = c("nCount_RNA"), pt.size = 0, group.by = "cell_type_1")  + scale_y_log10() + NoLegend() + geom_hline(yintercept = c(3500)) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)),
          VlnPlot(seurat_all, features = c("percent.mt"), pt.size = 0, group.by = "cell_type_1") + NoLegend() + geom_hline(yintercept = 10)+ theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)),
          ncol = 1)

gridExtra::grid.arrange(VlnPlot(seurat_all, features = c("nFeature_RNA"), pt.size = 0.0)  + scale_y_log10() + NoLegend() + geom_hline(yintercept = 2000),
          VlnPlot(seurat_all, features = c("nCount_RNA"), pt.size = 0.0) + NoLegend()  + scale_y_log10() + geom_hline(yintercept = c(3500)),
          VlnPlot(seurat_all, features = c("percent.mt"), pt.size = 0.0) + NoLegend() + geom_hline(yintercept = 10),
          ncol = 1)
```


```{r, fig.width = 6, fig.height = 4.5}
FeatureScatter(seurat_all, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "cell_type_1", pt.size = 1.5)
```

```{r, eval = FALSE}
saveRDS(seurat_all,"seurat_all_QC.rds")
```

```{r, eval = FALSE}
seurat_all <- readRDS("seurat_all_QC.rds")
```


```{r, eval = FALSE, warning = FALSE}
all.genes <- rownames(seurat_all)

seurat_all <- NormalizeData(seurat_all, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")
seurat_all <- ScaleData(seurat_all, features = all.genes, assay = "RNA")
seurat_all <- SCTransform(seurat_all, verbose = TRUE, return.only.var.genes = FALSE,assay = "RNA", n_genes = NULL, variable.features.n = NULL)
```



```{r, eval = FALSE}
saveRDS(seurat_all,"seurat_all_SCtransform.rds")
```

```{r, eval = FALSE}
seurat_all <- readRDS("seurat_all_SCtransform.rds")
```


```{r, eval = FALSE}
seurat_all <- RunPCA(seurat_all,
                       assay = "SCT",
                       features = all.genes,
                       npcs = 50, verbose = FALSE,
                       reduction.name = "pca")
```


```{r, eval = FALSE}
saveRDS(seurat_all,"seurat_all_PCA.rds")
```

```{r, eval = TRUE}
seurat_all <- readRDS("seurat_all_PCA.rds")
```

```{r}
ElbowPlot(seurat_all, ndims = 50)
```

```{r UMAP, eval = FALSE}
DefaultAssay(seurat_all) <- "SCT"
seurat_all <- RunUMAP(seurat_all, reduction = "pca", dims = 1:50, verbose = TRUE)
seurat_all <- FindNeighbors(seurat_all, reduction = "pca", dims = 1:50, verbose = TRUE)
seurat_all <- FindClusters(seurat_all, resolution = 0.5, verbose = TRUE)
```

```{r, eval = FALSE}
seurat_all$Sample <- factor(seurat_all$Sample,
                               levels = c("Ctrl_wt","Ctrl_w_a","Ctrl_AxD","OGD_wt","OGD_w_a","OGD_AxD"))
saveRDS(seurat_all,"seurat_all_UMAP.rds")
```

```{r, eval = TRUE}
seurat_all <- readRDS("seurat_all_UMAP.rds")
```

```{r}
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))[-4]
```


```{r pictures_plot1, messages = FALSE, fig.width = 6, fig.height = 6}
DimPlot(seurat_all, reduction = "umap", label = TRUE, cols = col_vector)
DimPlot(seurat_all, reduction = "umap", label = FALSE, group.by = "seurat_clusters", cols = col_vector)
DimPlot(seurat_all, reduction = "umap", label = FALSE, group.by = "Sample")
DimPlot(seurat_all, reduction = "umap", label = FALSE, split.by = "Sample", cols = col_vector)
DimPlot(seurat_all, reduction = "umap", label = TRUE, group.by = "cell_type_1", cols = col_vector)
```

```{r}
features_test <- c("SOX9","NFIB",      # ASTRO transformation
                   "CD44","GJA1","S100B", #General ASTRO
                   "GFAP","VIM",                      # Reactive ASTRO
                   "EPCAM","CDH1","CLDN7",  #Maturated iPSC from fibroblats
                   "SLC17A6","ISL1",
                   "MAP2","SYP","DCX","STMN1","TH", #Neuronal
                   "MALAT1","percent.mt"
                 )
```

```{r echo=FALSE, fig.height=6, fig.width = 6.2}
DotPlot(seurat_all, features = features_test, group.by = "cell_type_1") +
  theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5)) +
  geom_vline(xintercept = c(7.5,10.5), color = "darkgreen", linetype = 2) +
  geom_vline(xintercept = c(17.5), color = "darkgreen", linetype = 1)
```

```{r, fig.height=10, fig.width = 15}
FeaturePlot(seurat_all, features = features_test, min.cutoff = "q10", max.cutoff = "q90", ncol = 5, cols = c("gray","red"))
```

```{r}
ids <- unique(seurat_all@meta.data$Sample)
categories <- unique(seurat_all@meta.data$seurat_clusters)
counts <- matrix(nrow=length(ids), ncol=length(categories))
rownames(counts) <- ids
colnames(counts) <- categories

for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- seurat_all@meta.data %>%
      filter(Sample == ids[i], seurat_clusters == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}

counts <- counts[, order(as.numeric(colnames(counts)))]
counts <- counts/rowSums(counts)
counts_mm_1 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_1, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells') +
  xlab("")


counts <- t(counts)
counts <- counts/(rowSums(counts)/4)
counts <- t(counts)
counts_mm_2 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_2, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Relative proportion of cells') +
  xlab("")
```



```{r}
ggplot(data=seurat_all@meta.data, aes(seurat_clusters)) +
  geom_bar(aes(fill=cell_type_1), position="fill") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
  ylab('Proportion of cells')

ggplot(data=seurat_all@meta.data, aes(cell_type_1)) +
  geom_bar(aes(fill=seurat_clusters), position="fill") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
  ylab('Proportion of cells')

ggplot(data=seurat_all@meta.data, aes(seurat_clusters)) +
  geom_bar(aes(fill=Doublet), position="fill") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
  ylab('Proportion of cells')
```


```{r, fig.height=3, fig.width = 5}
VlnPlot(seurat_all, features = "percent.mt", pt.size = 0.2) +
   theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5)) + NoLegend()
VlnPlot(seurat_all, features = "percent.rib", pt.size = 0.2) +
   theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5)) + NoLegend()
VlnPlot(seurat_all, features = "nCount_RNA", pt.size = 0.2) +
   theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5)) + NoLegend() + scale_y_log10()
VlnPlot(seurat_all, features = "nFeature_RNA", pt.size = 0.2) +
   theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5)) + NoLegend() + scale_y_log10()
```





```{r message=FALSE, warning=FALSE}
seurat_ASTRO <- subset(seurat_all, subset = seurat_clusters %in% c("1","2","3","4","5","8","10","11"))
seurat_ASTRO$Cluster2 <- as.factor(Idents(seurat_ASTRO))
DefaultAssay(seurat_ASTRO) <- "RNA"
seurat_ASTRO$Sample <- factor(seurat_ASTRO$Sample, levels = c("Ctrl_wt","Ctrl_w_a","Ctrl_AxD","OGD_wt","OGD_w_a","OGD_AxD"))

all.genes <- rownames(seurat_ASTRO)
seurat_ASTRO <- NormalizeData(seurat_ASTRO, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")
seurat_ASTRO <- ScaleData(seurat_ASTRO, features = all.genes, assay = "RNA")
seurat_ASTRO <- SCTransform(seurat_ASTRO, verbose = TRUE, return.only.var.genes = FALSE,assay = "RNA", n_genes = NULL, variable.features.n = NULL)

seurat_ASTRO <- RunPCA(seurat_ASTRO,
                       assay = "SCT",
                       features = all.genes,
                       npcs = 50, verbose = FALSE,
                       reduction.name = "pca")
```

```{r, eval = FALSE}
saveRDS(seurat_ASTRO,"seurat_ASTRO_PCA.rds")
```

```{r, eval = FALSE}
seurat_ASTRO <- readRDS("seurat_ASTRO_PCA.rds")
```

```{r}
ElbowPlot(seurat_ASTRO, ndims = 50)
```



```{r message=FALSE, warning=FALSE}
seurat_ASTRO <- RunUMAP(seurat_ASTRO, reduction = "pca", dims = 1:38, verbose = TRUE, seed.use = 1902)
seurat_ASTRO <- FindNeighbors(seurat_ASTRO, reduction = "pca", dims = 1:38, verbose = TRUE)
seurat_ASTRO@meta.data <- seurat_ASTRO@meta.data[,substr(colnames(seurat_ASTRO@meta.data),1,11) != "SCT_snn_res"]
for(i in 1:10){
  seurat_ASTRO <- FindClusters(seurat_ASTRO, resolution = i*0.1, verbose = FALSE)
}
```



```{r eval=FALSE}
saveRDS(seurat_ASTRO,"seurat_ASTRO_clustree.rds")
```

```{r eval=TRUE}
seurat_ASTRO <- readRDS("seurat_ASTRO_clustree.rds")
```

```{r, fig.height=12, fig.width=8}
clustree(seurat_ASTRO, prefix = "SCT_snn_res.")
```


```{r}
seurat_ASTRO@meta.data <- seurat_ASTRO@meta.data[,substr(colnames(seurat_ASTRO@meta.data),1,11) != "SCT_snn_res"]
seurat_ASTRO <- FindClusters(seurat_ASTRO, resolution = 0.3, verbose = FALSE)
seurat_ASTRO$cell_type_2 <- Idents(seurat_ASTRO)
levels(seurat_ASTRO$cell_type_2) <- paste0("ASTRO_",as.numeric(levels(seurat_ASTRO$cell_type_2))+1)
```


```{r eval=FALSE}
saveRDS(seurat_ASTRO,"seurat_ASTRO_clusters.rds")
```

```{r eval=TRUE}
seurat_ASTRO <- readRDS("seurat_ASTRO_clusters.rds")
#write.csv(as.data.frame(subset(seurat_ASTRO, subset = cell_type_2 == "ASTRO_2", features = "GFAP")@assays$SCT@data), "seurat_ASTRO_SCT_data.csv")
#write.csv(as.data.frame(subset(seurat_ASTRO, subset = cell_type_2 == "ASTRO_2", features = "GFAP")@assays$SCT@counts), "seurat_ASTRO_SCT_counts.csv")
```


```{r}
seurat_ASTRO <- subset(seurat_ASTRO, subset = cell_type_2 != "ASTRO_6")
```


```{r eval=FALSE}
saveRDS(seurat_ASTRO,"seurat_ASTRO_subset.rds")
```

```{r eval=TRUE}
seurat_ASTRO <- readRDS("seurat_ASTRO_subset.rds")
```



```{r, fig.height=6, fig.width=12}
FeaturePlot(seurat_ASTRO, features = c("NFIB","SOX9","GLUL","VIM","CD44","GJA1","GFAP","S100B","EPCAM","CDH1","SLC1A3","C3"), min.cutoff = "q10", max.cutoff = "q90", order = TRUE)
FeaturePlot(seurat_ASTRO, features = c("NFIB","SOX9","GLUL","VIM","CD44","GJA1","GFAP","S100B","EPCAM","CDH1","SLC1A3","C3"), min.cutoff = "q10", max.cutoff = "q90", order = FALSE)
plot_density(seurat_ASTRO, features = c("NFIB","SOX9","GLUL","VIM","CD44","GJA1","GFAP","S100B","EPCAM","CDH1","SLC1A3","C3"), pal = "inferno")
DotPlot(seurat_ASTRO, features = c("GLUL","VIM","CD44","GJA1","GFAP","S100B","NFIB","SOX9","EPCAM","CDH1","SLC1A3")) + theme(axis.text.x=element_text(angle=90, hjust = 1, vjust = 0.5))
```

```{r pictures_plot5, messages = FALSE, fig.width = 9, fig.height = 6}
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Sample", cols = col_vector, pt.size = 1.2)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Condition", pt.size = 1.2)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Genotype", pt.size = 1.2)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Genotype_Neuro", pt.size = 1.2)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Genotype_Astro", pt.size = 1.2)

DimPlot(seurat_ASTRO, reduction = "umap", label = TRUE, group.by = "cell_type_1", cols = col_vector)
DimPlot(seurat_ASTRO, reduction = "umap", label = TRUE, group.by = "Cluster2")
DimPlot(seurat_ASTRO, reduction = "umap", label = TRUE, group.by = "cell_type_2", cols = col_vector)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "cell_type_2", split.by = "Sample", cols = col_vector)
DimPlot(seurat_ASTRO, reduction = "umap", label = FALSE, group.by = "Phase")
```



```{r}
ASTRO_monocle <- as.cell_data_set(seurat_ASTRO)
ASTRO_monocle <- cluster_cells(cds = ASTRO_monocle, reduction_method = "UMAP")
ASTRO_monocle <- learn_graph(ASTRO_monocle, use_partition = TRUE)
```


```{r}
ASTRO_monocle <- order_cells(cds = ASTRO_monocle, reduction_method = "UMAP")
```

```{r eval=FALSE}
saveRDS(ASTRO_monocle,"ASTRO_monocle.rds")
```

```{r eval=TRUE}
ASTRO_monocle <- readRDS("ASTRO_monocle.rds")
```


```{r}
plot_cells(
  cds = ASTRO_monocle,
  color_cells_by = "cell_type_2",
  show_trajectory_graph = TRUE
)
```

```{r}
library("reticulate")
use_condaenv("r-reticulate", required = TRUE)
scv <- import("scvelo")
scv$logging$print_version()
```

```{r}
samples<-read.table("../samples.csv",header=TRUE,sep=",")
```

```{r load_data, eval=FALSE, message = FALSE}
STAR_velocyto<-as.list(NULL)
for(i in 1:nrow(samples)) {
  folder <- paste0("../STAR/",samples$SampleName[i],"/",samples$SampleName[i],"Solo.out/Velocyto/raw/")
  
  cellbarcodes <- read.table(paste0(folder, "barcodes.tsv"))
  genenames <- read.table(paste0(folder, "genes.tsv"))
  dimnames <- as.list(NULL)
  dimnames[[1]] <- as.character(genenames$V2)
  dimnames[[2]] <- as.character(paste0(cellbarcodes$V1,"-",samples$SampleName[i]))
  
  STAR_velocyto[[i]]<-as.list(NULL)
  
  dims <- readr::read_delim(paste0(folder, "spliced.mtx"), delim = ' ', skip = 2, n_max = 1, col_names = FALSE, show_col_types = FALSE)
  t <- readr::read_delim(paste0(folder, "spliced.mtx"), delim = ' ', skip = 3, col_names = FALSE, show_col_types = FALSE)
  STAR_velocyto[[i]][["spliced"]] <- Matrix::sparseMatrix(i = t$X1, j = t$X2, x = t$X3, dims = c(dims$X1, dims$X2), dimnames = dimnames)
  cat(paste(samples$SampleName[i],"spliced\n"))
  
  dims <- readr::read_delim(paste0(folder, "unspliced.mtx"), delim = ' ', skip = 2, n_max = 1, col_names = FALSE, show_col_types = FALSE)
  t <- readr::read_delim(paste0(folder, "unspliced.mtx"), delim = ' ', skip = 3, col_names = FALSE, show_col_types = FALSE)
  STAR_velocyto[[i]][["unspliced"]] <- Matrix::sparseMatrix(i = t$X1, j = t$X2, x = t$X3, dims = c(dims$X1, dims$X2), dimnames = dimnames)
  cat(paste(samples$SampleName[i],"unspliced\n"))
}
```

```{r eval=FALSE}
saveRDS(STAR_velocyto,"STAR_velocyto.rds")
```

```{r eval=TRUE}
STAR_velocyto <- readRDS("STAR_velocyto.rds")
```


```{r eval=FALSE, warning=FALSE}
STAR_velocyto_merged <- as.list(NULL)
counts <- as.list(NULL)

STAR_velocyto_merged[["spliced"]] <- cbind(STAR_velocyto[[1]][["spliced"]],STAR_velocyto[[2]][["spliced"]],STAR_velocyto[[3]][["spliced"]],STAR_velocyto[[4]][["spliced"]],STAR_velocyto[[5]][["spliced"]],STAR_velocyto[[6]][["spliced"]])
counts[["spliced"]] <- STAR_velocyto_merged[["spliced"]][,colnames(STAR_velocyto_merged[["spliced"]]) %in% colnames(seurat_ASTRO)]

STAR_velocyto_merged[["unspliced"]] <- cbind(STAR_velocyto[[1]][["unspliced"]],STAR_velocyto[[2]][["unspliced"]],STAR_velocyto[[3]][["unspliced"]],STAR_velocyto[[4]][["unspliced"]],STAR_velocyto[[5]][["unspliced"]],STAR_velocyto[[6]][["unspliced"]])
counts[["unspliced"]] <- STAR_velocyto_merged[["unspliced"]][,colnames(STAR_velocyto_merged[["unspliced"]]) %in% colnames(seurat_ASTRO)]
```

```{r eval=FALSE}
rm(STAR_velocyto, STAR_velocyto_merged)
saveRDS(counts,"counts_ASTRO_velocyto.rds")
```

```{r eval=TRUE}
counts <- readRDS("counts_ASTRO_velocyto.rds")
```

```{r}
seurat_ASTRO[["spliced"]] <- CreateAssayObject(counts = counts[["spliced"]])
seurat_ASTRO[["unspliced"]] <- CreateAssayObject(counts = counts[["unspliced"]])

library("SeuratDisk")
DefaultAssay(seurat_ASTRO) <- "RNA"
SaveH5Seurat(seurat_ASTRO, filename = "seurat_ASTRO_velocyto.h5Seurat")
Convert("seurat_ASTRO_velocyto.h5Seurat", dest = "h5ad")
```

```{r}
library("reticulate")
use_condaenv("r-reticulate", required = TRUE)
scv <- import("scvelo")
```


```{r}
adata_ASTRO <- scv$read("seurat_ASTRO_velocyto.h5ad")
```

```{r}
scv$pp$filter_genes(adata_ASTRO)
scv$pp$moments(adata_ASTRO)
scv$tl$recover_dynamics(adata_ASTRO, n_jobs = 10)
#adata_ASTRO.__dict__['_raw'].__dict__['_var'] = adata_ASTRO.__dict__['_raw'].__dict__['_var'].rename(columns={'_index': 'features'})
adata_ASTRO$write("seurat_ASTRO_scVelo_graph.h5ad", compression = "gzip")
```

```{r}
#adata_ASTRO <- scv$read("seurat_ASTRO_scVelo_graph.h5ad")
```


```{r}
scv$tl$velocity(adata_ASTRO, mode='dynamical')
scv$tl$velocity_graph(adata_ASTRO)
adata_ASTRO$write("seurat_ASTRO_scVelo_graph2.h5ad", compression = "gzip")
```


```{r}
adata_ASTRO <- scv$read("seurat_ASTRO_scVelo_graph2.h5ad")
```



```{r}
scv$pl$velocity_embedding_stream(adata_ASTRO, basis='umap', color = "cell_type_2")
```

```{r}
topgenes <- adata_ASTRO$var["fit_likelihood"]
topgenes_val <- topgenes[,1]
names(topgenes_val) <- rownames(topgenes)
topgenes_val <- sort(topgenes_val, decreasing = TRUE)
head(topgenes_val, ntop = 10)
```

```{r}
scv$pl$scatter(adata_ASTRO, basis = names(topgenes_val)[1:10], ncols = 5L, frameon = FALSE, color = "cell_type_2")
```





```{r}
ids <- unique(seurat_ASTRO@meta.data$Sample)
categories <- unique(seurat_ASTRO@meta.data$cell_type_2)
counts <- matrix(nrow=length(ids), ncol=length(categories))
rownames(counts) <- ids
colnames(counts) <- categories

for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- seurat_ASTRO@meta.data %>%
      filter(Sample == ids[i], cell_type_2 == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}


counts <- counts[, order(colnames(counts))]
counts <- counts/rowSums(counts)
counts_mm_1 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_1, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells') +
  xlab("") + ggtitle("") + 
  theme(axis.text = element_text(family = "sans", size = 10, face = "plain", color = "black"),
        axis.title = element_text(family = "sans", size = 10, face = "plain"))


counts <- t(counts)
counts <- counts/(rowSums(counts)/4)
counts <- t(counts)
counts_mm_2 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_2, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Relative proportion of cells') +
  xlab("") + ggtitle("") + 
  theme(axis.text = element_text(family = "sans", size = 14, face = "bold", color = "black"),
        axis.title = element_text(family = "sans", size = 14, face = "bold"))
```



```{r , messages = FALSE, fig.width = 6, fig.height = 4}
#gi = c("EPCAM","CDH1","CLDN4","CLDN6","CLDN7")
gi = c("LRRC75A")
VlnPlot(seurat_ASTRO, features = gi, group.by = "cell_type_2") + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
VlnPlot(seurat_ASTRO, features = gi, group.by = "cell_type_2", split.by = "Condition") + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
FeaturePlot(seurat_ASTRO, features = gi, min.cutoff = "q20", max.cutoff = "q80")


VlnPlot(seurat_all, features = gi, group.by = "cell_type_1") + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
FeaturePlot(seurat_all, features = gi, min.cutoff = "q20", max.cutoff = "q80")
```



```{r pictures_plot6, messages = FALSE, fig.width = 9, fig.height = 4}
VlnPlot(seurat_ASTRO, features = "GFAP", group.by = "cell_type_2", split.by = "Sample") + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
FeaturePlot(seurat_ASTRO, features = "GFAP", min.cutoff = "q20", max.cutoff = "q80")
FeaturePlot(seurat_ASTRO, features = "S100B", min.cutoff = "q20", max.cutoff = "q80")
FeaturePlot(seurat_ASTRO, features = "NFIB", min.cutoff = "q20", max.cutoff = "q80")
FeaturePlot(seurat_ASTRO, features = "SOX9", min.cutoff = "q20", max.cutoff = "q80")
FeaturePlot(seurat_ASTRO, features = "VIM", min.cutoff = "q20", max.cutoff = "q80")
VlnPlot(seurat_ASTRO, features = "VIM", group.by = "cell_type_2")

FeaturePlot(seurat_ASTRO, features = "percent.mt", min.cutoff = "q1", max.cutoff = "q99", cols = c("gray","darkgreen","red"), pt.size = 1.0)
VlnPlot(seurat_ASTRO, features = "nFeature_RNA", group.by = "cell_type_2", pt.size = 0.01) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
VlnPlot(seurat_ASTRO, features = "nCount_RNA", group.by = "cell_type_2", pt.size = 0.01) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
VlnPlot(seurat_ASTRO, features = "percent.mt", group.by = "cell_type_2", pt.size = 0.01) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
VlnPlot(seurat_ASTRO, features = "percent.rib", group.by = "cell_type_2", pt.size = 0.01) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
VlnPlot(seurat_ASTRO, features = "MALAT1", group.by = "cell_type_2", pt.size = 0.01) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```




```{r}
seurat_ASTRO@meta.data <- seurat_ASTRO@meta.data[,substr(colnames(seurat_ASTRO@meta.data),1,11) != "SCT_snn_res"]
seurat_ASTRO$cell_type_3 <- seurat_ASTRO$cell_type_2

levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_1"] <- "iAs"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_2"] <- "iAs"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_3"] <- "intermediate"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_4"] <- "iPSCs"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_5"] <- "iPSCs"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_7"] <- "iPSCs"
levels(seurat_ASTRO$cell_type_3)[levels(seurat_ASTRO$cell_type_3) == "ASTRO_8"] <- "AxD_WT"
```

```{r eval=FALSE}
saveRDS(seurat_ASTRO,"seurat_ASTRO_merged_clusters.rds")
```

```{r eval=TRUE}
seurat_ASTRO <- readRDS("seurat_ASTRO_merged_clusters.rds")
```

```{r pictures_plot5, messages = FALSE, fig.width = 9, fig.height = 6}
DimPlot(seurat_ASTRO, reduction = "umap", label = TRUE, group.by = "cell_type_3", cols = col_vector)
```

```{r}
ids <- unique(seurat_ASTRO@meta.data$Sample)
categories <- unique(seurat_ASTRO@meta.data$cell_type_3)
counts <- matrix(nrow=length(ids), ncol=length(categories))
rownames(counts) <- ids
colnames(counts) <- categories

for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- seurat_ASTRO@meta.data %>%
      filter(Sample == ids[i], cell_type_3 == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}


counts <- counts[, order(colnames(counts))]
counts <- counts/rowSums(counts)
counts_mm_1 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_1, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells') +
  xlab("") + ggtitle("") + 
  theme(axis.text = element_text(family = "sans", size = 10, face = "plain", color = "black"),
        axis.title = element_text(family = "sans", size = 10, face = "plain"))
```






```{r}
markers_iPSCs <- FindMarkers(seurat_ASTRO,
                                           ident.1 = c("iPSCs"),
                                           ident.2 = c("iAs"),
                                           group.by = "cell_type_3",
                                           min.pct = 0.25,
                                           only.pos = FALSE,
                                           verbose = TRUE,
                                           assay = "SCT",
                                           test.use = "wilcox")

#markers_iPSCs <- FindMarkers(subset(seurat_ASTRO, subset = cell_type_2 %in% c("ASTRO_2") & Genotype_Astro %in% "AxD"),
#                                           ident.1 = c("OGD"),
#                                           ident.2 = c("Ctrl"),
#                                           group.by = "Condition",
#                                           min.pct = 0.25,
#                                           only.pos = FALSE,
#                                           verbose = TRUE,
#                                           assay = "SCT",
#                                           test.use = "wilcox")



markers_intermediate_iPSCs <- FindMarkers(seurat_ASTRO,
                                           ident.1 = c("intermediate"),
                                           ident.2 = c("iPSCs"),
                                           group.by = "cell_type_3",
                                           min.pct = 0.25,
                                           only.pos = FALSE,
                                           verbose = TRUE,
                                           assay = "SCT",
                                           test.use = "wilcox")

markers_intermediate_iAs <- FindMarkers(seurat_ASTRO,
                                           ident.1 = c("intermediate"),
                                           ident.2 = c("iAs"),
                                           group.by = "cell_type_3",
                                           min.pct = 0.25,
                                           only.pos = FALSE,
                                           verbose = TRUE,
                                           assay = "SCT",
                                           test.use = "wilcox")

markers_AxD_WT <- FindMarkers(seurat_ASTRO,
                                           ident.1 = c("AxD_WT"),
                                           ident.2 = NULL,
                                           group.by = "cell_type_3",
                                           min.pct = 0.25,
                                           only.pos = FALSE,
                                           verbose = TRUE,
                                           assay = "SCT",
                                           test.use = "wilcox")
```


```{r}
#datatable(markers_IAA)
write.csv(markers_iPSCs,"markers_iPSCs.csv")
write.csv(markers_intermediate_iPSCs,"markers_intermediate_iPSCs.csv")
write.csv(markers_intermediate_iAs,"markers_intermediate_iAs.csv")
write.csv(markers_AxD_WT,"markers_AxD_WT.csv")
```

```{r}
reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
list_gene_names_down <- rownames(markers_iPSCs[markers_iPSCs$avg_log2FC > 0.6 & markers_iPSCs$p_val_adj < 1E-1,])
list_gene_names_up <- c(rownames(markers_iPSCs[markers_iPSCs$avg_log2FC < -0.6 & markers_iPSCs$p_val_adj < 1E-1,]))


topT <- markers_iPSCs
topT <- topT[order(-topT$p_val_adj),]
topT$p_val_adj[topT$p_val_adj < 1e-300] <- 1e-300
ggplot(aes(x = avg_log2FC, y = p_val_adj), data = topT) +
  labs(y = expression(Log[10]*"(adjusted p-value)"),
       x = expression(Log[2]*"(fold change)"),
       #title = expression("iPSCs vs iAs")) +
       title = expression("ASTRO_2 AxD OGD vs Ctrl")) +
    #xlim(-1.5,1.5) +
  geom_point(color = ifelse(topT$p_val_adj < 0.1 & !is.na(topT$p_val_adj) & abs(topT$avg_log2FC) > 0.6, "red","black")) +
  scale_y_continuous(trans = reverselog_trans(10)) +
  theme(legend.position = "none") +

   # downregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),], pch = 21, fill = NA, cex = 3, colour = "dodgerblue") +
  
  # upregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),], pch = 21, fill = NA, cex = 3, colour = "darkred") +
  
  # add label names
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),])),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   color = "darkred",
                   size = 4,
  #                 xlim = c(-0.5,0.25),
  #                 ylim = c(7.5,11.5),
                   max.iter = 20000) +
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),])),
                   box.padding   = 0.01, 
                   point.padding = 0.5,
                   color = "dodgerblue",
                   size = 4,
  #                 xlim = c(0.25,1),
  #                 ylim = c(0.1,7.5),
                   max.iter = 20000) +
 theme(panel.border = element_blank(),
      panel.grid.major = element_line(colour = "#d3d3d3"),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(colour = "black", fill = NA),
      plot.title = element_text(size = 10, family = "Arial", face = "bold", hjust = 0.5),
      text=element_text(family = "Arial"),
      axis.title = element_text(colour="black", size = 10, face="bold"),
      axis.text.x = element_text(colour="black", size = 10, face="plain"),
      axis.text.y = element_text(colour="black", size = 10, face="plain"),
      axis.line = element_line(size=0.5, colour = "black"))
```


```{r}
subfunction1 <- function(test){
  gene_list <- test$avg_log2FC * (-log10(test$p_val))
  names(gene_list) <- rownames(test)
  gene_list <- na.omit(gene_list)
  keep <- is.finite(gene_list)
  gene_list <- gene_list[keep]
  gene_list <- sort(gene_list, decreasing = TRUE)
  return(gene_list)
}



genes <- markers_iPSCs
genes$p_val <- genes$p_val_adj
genes$p_val[genes$p_val < 1e-300] <- 1e-300
gene.list <- subfunction1(genes)

# Run GSEA.

gse <- gseGO(geneList=gene.list, 
             ont ="BP", 
             keyType = "SYMBOL", 
             #nPerm = 50000,
             nPermSimple = 50000,
             minGSSize = 5, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db, 
             pAdjustMethod = "BH",
             by = "fgsea")
```

```{r}
if (dim(gse)[1] != 0){
  gse <- gse[gse@result$setSize >= 5, asis = T]
  gse <- gse[gse@result$p.adjust <= 0.1, asis = T]
}
if (dim(gse)[1] != 0){
  n.genes.list <- c()
  for (g in 1:length(gse@result[["core_enrichment"]])){
    list.genes <- unlist(strsplit(gse@result[["core_enrichment"]][g], split = "/"))
    n.genes <- length(list.genes)
    n.genes.list <- append(n.genes.list, n.genes)
  }
  gse@result[["n.genes"]] <- n.genes.list
  gse <- gse[gse@result$n.genes >= 3, asis = T]
}
dotplot(gse, showCategory = 12, 
             title = "GSEA", split=".sign") + facet_grid(.~.sign)
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
list_gene_names_down <- rownames(markers_AxD_WT[markers_AxD_WT$avg_log2FC > 1 & markers_AxD_WT$p_val_adj < 1E-3,])
list_gene_names_up <- rownames(markers_AxD_WT[markers_AxD_WT$avg_log2FC < -1 & markers_AxD_WT$p_val_adj < 1E-3,])


topT <- markers_AxD_WT
topT <- topT[order(-topT$p_val_adj),]
topT$p_val_adj[topT$p_val_adj < 1e-300] <- 1e-300
ggplot(aes(x = avg_log2FC, y = p_val_adj), data = topT) +
  labs(y = expression(Log[10]*"(adjusted p-value)"),
       x = expression(Log[2]*"(fold change)"),
       title = expression("markers AxD_WT vs other")) +
    #xlim(-1.5,1.5) +
  geom_point(color = ifelse(topT$p_val_adj < 0.1 & !is.na(topT$p_val_adj) & abs(topT$avg_log2FC) > 0.6, "red","black")) +
  scale_y_continuous(trans = reverselog_trans(10)) +
  theme(legend.position = "none") +

   # downregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),], pch = 21, fill = NA, cex = 3, colour = "dodgerblue") +
  
  # upregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),], pch = 21, fill = NA, cex = 3, colour = "darkred") +
  
  # add label names
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),])),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   color = "darkred",
                   size = 4,
  #                 xlim = c(-0.5,0.25),
  #                 ylim = c(7.5,11.5),
                   max.iter = 20000) +
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),])),
                   box.padding   = 0.01, 
                   point.padding = 0.5,
                   color = "dodgerblue",
                   size = 4,
  #                 xlim = c(0.25,1),
  #                 ylim = c(0.1,7.5),
                   max.iter = 20000) +
 theme(panel.border = element_blank(),
      panel.grid.major = element_line(colour = "#d3d3d3"),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(colour = "black", fill = NA),
      plot.title = element_text(size = 10, family = "Arial", face = "bold", hjust = 0.5),
      text=element_text(family = "Arial"),
      axis.title = element_text(colour="black", size = 10, face="bold"),
      axis.text.x = element_text(colour="black", size = 10, face="plain"),
      axis.text.y = element_text(colour="black", size = 10, face="plain"),
      axis.line = element_line(size=0.5, colour = "black"))
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
list_gene_names_down <- rownames(markers_intermediate_iPSCs[markers_intermediate_iPSCs$avg_log2FC > 1 & markers_intermediate_iPSCs$p_val_adj < 1E-3,])
list_gene_names_up <- c(rownames(markers_intermediate_iPSCs[markers_intermediate_iPSCs$avg_log2FC < -1 & markers_intermediate_iPSCs$p_val_adj < 1E-3,]))


topT <- markers_intermediate_iPSCs
topT <- topT[order(-topT$p_val_adj),]
topT$p_val_adj[topT$p_val_adj < 1e-300] <- 1e-300
ggplot(aes(x = avg_log2FC, y = p_val_adj), data = topT) +
  labs(y = expression(Log[10]*"(adjusted p-value)"),
       x = expression(Log[2]*"(fold change)"),
       title = expression("intermediate vs iPSCs")) +
    #xlim(-1.5,1.5) +
  geom_point(color = ifelse(topT$p_val_adj < 0.1 & !is.na(topT$p_val_adj) & abs(topT$avg_log2FC) > 0.6, "red","black")) +
  scale_y_continuous(trans = reverselog_trans(10)) +
  theme(legend.position = "none") +

   # downregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),], pch = 21, fill = NA, cex = 3, colour = "dodgerblue") +
  
  # upregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),], pch = 21, fill = NA, cex = 3, colour = "darkred") +
  
  # add label names
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),])),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   color = "darkred",
                   size = 4,
  #                 xlim = c(-0.5,0.25),
  #                 ylim = c(7.5,11.5),
                   max.iter = 20000) +
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),])),
                   box.padding   = 0.01, 
                   point.padding = 0.5,
                   color = "dodgerblue",
                   size = 4,
  #                 xlim = c(0.25,1),
  #                 ylim = c(0.1,7.5),
                   max.iter = 20000) +
 theme(panel.border = element_blank(),
      panel.grid.major = element_line(colour = "#d3d3d3"),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(colour = "black", fill = NA),
      plot.title = element_text(size = 10, family = "Arial", face = "bold", hjust = 0.5),
      text=element_text(family = "Arial"),
      axis.title = element_text(colour="black", size = 10, face="bold"),
      axis.text.x = element_text(colour="black", size = 10, face="plain"),
      axis.text.y = element_text(colour="black", size = 10, face="plain"),
      axis.line = element_line(size=0.5, colour = "black"))
```


```{r echo=FALSE, message=FALSE, warning=FALSE}
list_gene_names_down <- rownames(markers_intermediate_iAs[markers_intermediate_iAs$avg_log2FC > 1 & markers_intermediate_iAs$p_val_adj < 1E-3,])
list_gene_names_up <- c(rownames(markers_intermediate_iAs[markers_intermediate_iAs$avg_log2FC < -1 & markers_intermediate_iAs$p_val_adj < 1E-3,]))


topT <- markers_intermediate_iAs
topT <- topT[order(-topT$p_val_adj),]
topT$p_val_adj[topT$p_val_adj < 1e-300] <- 1e-300
ggplot(aes(x = avg_log2FC, y = p_val_adj), data = topT) +
  labs(y = expression(Log[10]*"(adjusted p-value)"),
       x = expression(Log[2]*"(fold change)"),
       title = expression("intermediate vs iAs")) +
    #xlim(-1.5,1.5) +
  geom_point(color = ifelse(topT$p_val_adj < 0.1 & !is.na(topT$p_val_adj) & abs(topT$avg_log2FC) > 0.6, "red","black")) +
  scale_y_continuous(trans = reverselog_trans(10)) +
  theme(legend.position = "none") +

   # downregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),], pch = 21, fill = NA, cex = 3, colour = "dodgerblue") +
  
  # upregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),], pch = 21, fill = NA, cex = 3, colour = "darkred") +
  
  # add label names
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),])),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   color = "darkred",
                   size = 4,
  #                 xlim = c(-0.5,0.25),
  #                 ylim = c(7.5,11.5),
                   max.iter = 20000) +
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),])),
                   box.padding   = 0.01, 
                   point.padding = 0.5,
                   color = "dodgerblue",
                   size = 4,
  #                 xlim = c(0.25,1),
  #                 ylim = c(0.1,7.5),
                   max.iter = 20000) +
 theme(panel.border = element_blank(),
      panel.grid.major = element_line(colour = "#d3d3d3"),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(colour = "black", fill = NA),
      plot.title = element_text(size = 10, family = "Arial", face = "bold", hjust = 0.5),
      text=element_text(family = "Arial"),
      axis.title = element_text(colour="black", size = 10, face="bold"),
      axis.text.x = element_text(colour="black", size = 10, face="plain"),
      axis.text.y = element_text(colour="black", size = 10, face="plain"),
      axis.line = element_line(size=0.5, colour = "black"))
```




```{r, fig.height=6, fig.width=12}
plot_density(seurat_ASTRO, features = c("MAP1B","GAP43","SYT4","LRRC75A","STMN2","PEG10"), pal = "inferno")
plot_density(seurat_ASTRO, features = c("WFDC2","KRT19","FN1","C3","APP","F3","MMP7","NEAT1","MALAT1"), pal = "inferno")
plot_density(seurat_ASTRO, features = c("C1QL1","SERPINF1","COL2A1","MT2A","S100B","CRYAB","DNAJC1","HSPB6","FABP7"), pal = "inferno")
plot_density(seurat_ASTRO, features = c("COL1A2","COL1A1","COL3A1","IGFBP7","SFRP2","SFRP1","ITGA11","TGFBI","RARRES2","ACTA2","MMP2","CCDC80"), pal = "inferno")
```






```{r message=FALSE, warning=FALSE}
seurat_NEURO <- subset(seurat_all, subset = seurat_clusters %in% c("0","9","6","12"))
seurat_NEURO$Cluster2 <- as.factor(Idents(seurat_NEURO))
DefaultAssay(seurat_NEURO) <- "RNA"

all.genes <- rownames(seurat_NEURO)
seurat_NEURO <- NormalizeData(seurat_NEURO, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")
seurat_NEURO <- ScaleData(seurat_NEURO, features = all.genes, assay = "RNA")
seurat_NEURO <- SCTransform(seurat_NEURO, verbose = TRUE, return.only.var.genes = FALSE,assay = "RNA", n_genes = NULL, variable.features.n = NULL)

seurat_NEURO <- RunPCA(seurat_NEURO,
                       assay = "SCT",
                       features = all.genes,
                       npcs = 50, verbose = FALSE,
                       reduction.name = "pca")
```

```{r, eval = FALSE}
saveRDS(seurat_NEURO,"seurat_NEURO_PCA.rds")
```

```{r, eval = FALSE}
seurat_NEURO <- readRDS("seurat_NEURO_PCA.rds")
```

```{r}
ElbowPlot(seurat_NEURO, ndims = 50)
```



```{r message=FALSE, warning=FALSE}
seurat_NEURO <- RunUMAP(seurat_NEURO, reduction = "pca", dims = 1:47, verbose = TRUE, seed.use = 1902)
seurat_NEURO <- FindNeighbors(seurat_NEURO, reduction = "pca", dims = 1:47, verbose = TRUE)
seurat_NEURO@meta.data <- seurat_NEURO@meta.data[,substr(colnames(seurat_NEURO@meta.data),1,11) != "SCT_snn_res"]
for(i in 1:10){
  seurat_NEURO <- FindClusters(seurat_NEURO, resolution = i*0.1, verbose = FALSE)
}
```



```{r eval=FALSE}
saveRDS(seurat_NEURO,"seurat_NEURO_clustree.rds")
```

```{r eval=TRUE}
seurat_NEURO <- readRDS("seurat_NEURO_clustree.rds")
```

```{r, fig.height=12, fig.width=8}
clustree(seurat_NEURO, prefix = "SCT_snn_res.")
```


```{r}
seurat_NEURO@meta.data <- seurat_NEURO@meta.data[,substr(colnames(seurat_NEURO@meta.data),1,11) != "SCT_snn_res"]
seurat_NEURO <- FindClusters(seurat_NEURO, resolution = 0.3, verbose = FALSE)
seurat_NEURO$cell_type_2 <- Idents(seurat_NEURO)
levels(seurat_NEURO$cell_type_2) <- paste0("NEURO_",as.numeric(levels(seurat_NEURO$cell_type_2))+1)
```


```{r eval=FALSE}
saveRDS(seurat_NEURO,"seurat_NEURO_clusters.rds")
```

```{r eval=TRUE}
seurat_NEURO <- readRDS("seurat_NEURO_clusters.rds")
```


```{r pictures_plot5, messages = FALSE, fig.width = 9, fig.height = 6}
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Sample", cols = col_vector, pt.size = 1.2)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Condition", pt.size = 1.2)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Genotype", pt.size = 1.2)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Genotype_Neuro", pt.size = 1.2)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Genotype_Astro", pt.size = 1.2)

DimPlot(seurat_NEURO, reduction = "umap", label = TRUE, group.by = "cell_type_1", cols = col_vector)
DimPlot(seurat_NEURO, reduction = "umap", label = TRUE, group.by = "Cluster2")
DimPlot(seurat_NEURO, reduction = "umap", label = TRUE, group.by = "cell_type_2", cols = col_vector)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "cell_type_2", split.by = "Sample", cols = col_vector)
DimPlot(seurat_NEURO, reduction = "umap", label = FALSE, group.by = "Phase")
```



```{r}
ids <- unique(seurat_NEURO@meta.data$Sample)
categories <- unique(seurat_NEURO@meta.data$cell_type_2)
counts <- matrix(nrow=length(ids), ncol=length(categories))
rownames(counts) <- ids
colnames(counts) <- categories

for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- seurat_NEURO@meta.data %>%
      filter(Sample == ids[i], cell_type_2 == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}

counts <- counts[, order(colnames(counts))]
counts <- counts/rowSums(counts)
counts_mm_1 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_1, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells') +
  xlab("") + ggtitle("") + 
  theme(axis.text = element_text(family = "sans", size = 10, face = "plain", color = "black"),
        axis.title = element_text(family = "sans", size = 10, face = "plain"))


counts <- t(counts)
counts <- counts/(rowSums(counts)/4)
counts <- t(counts)
counts_mm_2 <- melt(counts, id = "rownames")

ggplot(data = counts_mm_2, aes(x = as.factor(Var2), y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Relative proportion of cells') +
  xlab("") + ggtitle("") + 
  theme(axis.text = element_text(family = "sans", size = 10, face = "plain", color = "black"),
        axis.title = element_text(family = "sans", size = 10, face = "plain"))
```



```{r}
seurat_NEURO_subset <- subset(seurat_NEURO, subset = cell_type_2 %in% c("NEURO_1","NEURO_2","NEURO_3","NEURO_4"))
```

```{r}
DimPlot(seurat_NEURO_subset, reduction = "umap", label = TRUE, group.by = "cell_type_2", cols = col_vector)
```

```{r}
markers <- FindMarkers(seurat_NEURO_subset,
                                           ident.1 = c("AxD_AxD"),
                                           ident.2 = c("WT_AxD","WT_WT"),
                                           group.by = "Genotype",
                                           min.pct = 0.25,
                                           only.pos = FALSE,
                                           verbose = TRUE,
                                           assay = "SCT",
                                           test.use = "wilcox")
```

Ctrl_AxD Ctrl_w_a  Ctrl_wt  OGD_AxD  OGD_w_a   OGD_wt 

```{r}
#datatable(markers_IAA)
#write.csv(markers_iPSCs,"markers_iPSCs.csv")
```

```{r}
reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
list_gene_names_down <- rownames(markers[markers$avg_log2FC > 0.6 & markers$p_val_adj < 1E-3,])
list_gene_names_up <- rownames(markers[markers$avg_log2FC < -0.6 & markers$p_val_adj < 1E-3,])


topT <- markers
topT <- topT[order(-topT$p_val_adj),]
topT$p_val_adj[topT$p_val_adj < 1e-300] <- 1e-300
ggplot(aes(x = avg_log2FC, y = p_val_adj), data = topT) +
  labs(y = expression(Log[10]*"(adjusted p-value)"),
       x = expression(Log[2]*"(fold change)"),
       title = expression("AxD_AxD vs AxD_WT, WT_WT")) +
    #xlim(-1.5,1.5) +
  geom_point(color = ifelse(topT$p_val_adj < 0.1 & !is.na(topT$p_val_adj) & abs(topT$avg_log2FC) > 0.6, "red","black")) +
  scale_y_continuous(trans = reverselog_trans(10)) +
  theme(legend.position = "none") +

   # downregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),], pch = 21, fill = NA, cex = 3, colour = "dodgerblue") +
  
  # upregulated
  geom_point(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),], pch = 21, fill = NA, cex = 3, colour = "darkred") +
  
  # add label names
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_up),])),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   color = "darkred",
                   size = 4,
  #                 xlim = c(-0.5,0.25),
  #                 ylim = c(7.5,11.5),
                   max.iter = 20000) +
  geom_label_repel(data = topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),],
                   aes(label = rownames(topT[toupper(rownames(topT)) %in% toupper(list_gene_names_down),])),
                   box.padding   = 0.01, 
                   point.padding = 0.5,
                   color = "dodgerblue",
                   size = 4,
  #                 xlim = c(0.25,1),
  #                 ylim = c(0.1,7.5),
                   max.iter = 20000) +
 theme(panel.border = element_blank(),
      panel.grid.major = element_line(colour = "#d3d3d3"),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(colour = "black", fill = NA),
      plot.title = element_text(size = 10, family = "Arial", face = "bold", hjust = 0.5),
      text=element_text(family = "Arial"),
      axis.title = element_text(colour="black", size = 10, face="bold"),
      axis.text.x = element_text(colour="black", size = 10, face="plain"),
      axis.text.y = element_text(colour="black", size = 10, face="plain"),
      axis.line = element_line(size=0.5, colour = "black"))
```

```{r}
VlnPlot(seurat_NEURO, feature = "PTH2", group.by = "Sample")
VlnPlot(seurat_NEURO, feature = "UCN", group.by = "Sample")
VlnPlot(seurat_NEURO, feature = "SCG5", group.by = "Sample")
VlnPlot(seurat_NEURO, feature = "ATP5E", group.by = "Sample")
VlnPlot(seurat_NEURO, feature = "MGP", group.by = "Sample")
VlnPlot(seurat_NEURO, feature = "SOX4", group.by = "Sample")
```

```{r}

```



```{r}
subfunction1 <- function(test){
  gene_list <- test$avg_log2FC * (-log10(test$p_val))
  names(gene_list) <- rownames(test)
  gene_list <- na.omit(gene_list)
  keep <- is.finite(gene_list)
  gene_list <- gene_list[keep]
  gene_list <- sort(gene_list, decreasing = TRUE)
  return(gene_list)
}



genes <- markers
genes$p_val <- genes$p_val_adj
genes$p_val[genes$p_val < 1e-300] <- 1e-300
gene.list <- subfunction1(genes)

# Run GSEA.

gse <- gseGO(geneList=gene.list, 
             ont ="BP", 
             keyType = "SYMBOL", 
             #nPerm = 50000,
             nPermSimple = 50000,
             minGSSize = 5, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db, 
             pAdjustMethod = "BH",
             by = "fgsea")
```


```{r, fig.height=6, fig.width=12}
plot_density(seurat_NEURO, features = c("PTH2","SYT4","SCG5","UCN","CST3","NNAT","TRH","TAC1","PCSK1"), pal = "inferno")
plot_density(seurat_NEURO, features = c("ATP5E","BTF3","MGP","KRT19","WFDC2","S100A11","SOX4","S100A10","NTS"), pal = "inferno")
```


